### DISCUSSION TITLE
Bandgap Prediction using MACE

### DISCUSSION THREAD

**Discussion by amannsonii3** [ACEsuit/mace#917] [Q&A] [answered]
**Detailed Description**
Bandgap, unlike energy, is an intrinsic property of the system, unchanging with the size of the system (e.g. constructing supercells of the system, unlike energy, would yield the same constant bandgap). 
I was wondering if we were to replicate certain energy functions in MACE, but using mean or scatter_mean instead of sum or scatter_sum, we should be able to train a model that can predict bandgaps.

**Features Request**
New functions in MACECalculator for bandgap calculator and replication of certain energy functions in MACE modified for bandgaps.

**Modifications I tried**
I tried editing def forward within MACE and ScaleShiftMACE class in the models.py file. I also tried modifying the weighted_mean_squared_error_energy function in loss.py 

1. mace/modules/loss.py: 
```
return torch.mean(
        configs_weight
        * configs_energy_weight
        * torch.square((ref["energy"] - pred["energy"])) # Does it matter if we keep the division by num_atoms or not?
    )
```
2. mace/modules/models.py
```
e0 = scatter_mean(
            src=node_e0, index=data["batch"], dim=-1, dim_size=num_graphs
        )
```
```
energy = scatter_mean(
                src=node_energies, index=data["batch"], dim=-1, dim_size=num_graphs
            )  # [n_graphs,]
```

``` 
total_energy = torch.mean(contributions, dim=-1) # [n_graphs,] 
```

**Additional context**
P.S. I am not very comfortable with the MACE architecture like the developers and contributors. It would be great to have assistance in building this feature. Please let me know if such a feature already exists and I might have missed it

**Comment by tisabe** [comment]
I would also be interested in this feature, since often band gap prediction and structure relaxation appear together in a workflow. Ideally one would just write a new loss function, that is only "energy per atom" loss. But in the dataset, you would have to label the bandgap values as energies. This kind of loss would also be applicable to many other invariant scalar functions.

**Comment by ilyes319** [comment, marked as answer]
Hey @amannsonii3,

As said, one needs to output a global intensive property. There are many ways to do so, using scatter_mean is one.
Here is one approach i have done that worked well:

```
        inter_e = scatter_mean(
            src=node_energies,
            index=data.batch.unsqueeze(-1),
            dim=0,
            dim_size=data.num_graphs,
        )  # [n_graphs,16]
        inter_std = scatter_std(
            src=node_energies,
            index=data.batch.unsqueeze(-1),
            dim=0,
            dim_size=data.num_graphs,
        )  # [n_graphs,16]
        inter_sum = scatter_sum(
            src=node_energies,
            index=data.batch.unsqueeze(-1),
            dim=0,
            dim_size=data.num_graphs,
        )  # [n_graphs,16]

        inter_e = inter_e[:, :, None]
        inter_std = inter_std[:, :, None]
        inter_sum = inter_sum[:, :, None]

        momentums = self.mom_mapper(torch.cat([inter_e, inter_std, inter_sum], dim=2))
        momentums = momentums.reshape(momentums.shape[0], 1, momentums.shape[1])
        att_momentums, _ = self.mom_attn(momentums, momentums, momentums)
        momentums = momentums + att_momentums
        momentums = momentums[:, 0, :]
        probabilities = self.fc(momentums)
```

with 

```
        # Final fully connected layer
        input_dim = self.readouts[0].irreps_out.num_irreps

        self.mom_mapper = torch.nn.Sequential(
            torch.nn.Linear(3, 1),
            torch.nn.GELU(),
            torch.nn.Dropout(0.01),
        )

        mid_dim = MLP_irreps.num_irreps
        self.mom_attn = torch.nn.MultiheadAttention(
            input_dim, 8, 0.05, batch_first=True
        )

        self.fc = torch.nn.Sequential(
            torch.nn.Linear(input_dim, mid_dim),
            torch.nn.GELU(),
            torch.nn.Dropout(0.01),
            torch.nn.Linear(mid_dim, 2),
        )
```

**Reply by bismuthadams1** [reply]
Hi Ilyes, thanks this reply. To jump on the discussion, I'm following the same process here but with the final layer I'm outputting a binary classifier (hence the final output dim is 1):

```
class TransformerGraphReadoutBlock(torch.nn.Module):
    def __init__(
        self,
        irreps_out: o3.Irreps,
        MLP_irreps: o3.Irreps,
        cueq_config: Optional[CuEquivarianceConfig] = None,
    ):
        super().__init__()
        
        self.pool_norm = torch.nn.LayerNorm(3)

        input_dim = irreps_out.dim # input dimension for the MLP

        self.mapper = torch.nn.Sequential(
            torch.nn.Linear(3,1),
            torch.nn.GELU(),
            torch.nn.Dropout(0.1),
        ) # out dimension is 1, input dimension is 3 (inter_e, inter_std, inter_sum)

        mid_dim = MLP_irreps.num_irreps
        self.attn = torch.nn.MultiheadAttention(
            input_dim, 8, 0.05, batch_first=True
        )

        self.fc = torch.nn.Sequential(
            torch.nn.Linear(input_dim, mid_dim),
            torch.nn.GELU(),
            torch.nn.Dropout(0.1),
            torch.nn.Linear(mid_dim, 1),
        )

    def forward(self, x: tuple[torch.Tensor]) -> torch.Tensor:
        inter_e, inter_std, inter_sum  = x

        momentums = self.mapper(torch.cat([inter_e, inter_std, inter_sum], dim=2)) # [n_graphs, 16],from input 

        momentums = momentums.reshape(momentums.shape[0], 1, momentums.shape[1])  # [n_graphs,1,16]
        # logging.info(f"momentums into attention: {momentums}")

        att_momentums, _ = self.attn(
            momentums, momentums, momentums
        )  # [n_graphs,1,16]. attn_output, attn_output_weights = multihead_attn(query, key, value)
        momentums = momentums + att_momentums  # [n_graphs,1,16] add the attention output to the momentums
        momentums = momentums[:, 0, :]  # [n_graphs,16] remove the second dimension
        graph_logits = self.fc(momentums).squeeze(-1)  # [n_graphs,16] apply the fully connected layer

        return graph_logits
```

Where I'm using a `torch.nn.functional.binary_cross_entropy_with_logits` loss function. I've noticed I'm getting vanishing gradients (the classifier hits an accuracy wall despite size of model), despite not using a sigmoid directly in the readout block. I've only added a readout function (your suggested transformer). However, I'm wondering if its necessary for each of the layers to have a readout function, this is what I have:

```
  def forward(
     self,
        data: Dict[str, torch.Tensor],
        training: bool = False,  # pylint: disable=W0613
        compute_force: bool = False,
        compute_virials: bool = False,
        compute_stress: bool = False,
        compute_displacement: bool = False,
        compute_edge_forces: bool = False,  # pylint: disable=W0613
        compute_atomic_stresses: bool = False,  # pylint: disable=W0613,
        compute_coupling_class: bool = True,  # pylint: disable=W0613
    ) -> Dict[str, Optional[torch.Tensor]]:
        assert compute_force is False
        assert compute_virials is False
        assert compute_stress is False
        assert compute_displacement is False
        assert compute_coupling_class is True
  
        # Setup
        data["node_attrs"].requires_grad_(True)
        data["positions"].requires_grad_(True)
        num_graphs = data["ptr"].numel() - 1
  
        # Embeddings
        node_feats = self.node_embedding(data["node_attrs"])
  
        logging.info('node feats prior to run')
        logging.info(node_feats)
  
        vectors, lengths = get_edge_vectors_and_lengths(
            positions=data["positions"],
            edge_index=data["edge_index"],
            shifts=data["shifts"],
        )
        edge_attrs = self.spherical_harmonics(vectors)
        edge_feats, cutoff = self.radial_embedding(
            lengths, data["node_attrs"], data["edge_index"], self.atomic_numbers
        )
  
       node_feats_total = []
       for interaction, product in zip(
            self.interactions, self.products
        ):
            node_feats, sc = interaction(
                node_attrs=data["node_attrs"],
                node_feats=node_feats,
                edge_attrs=edge_attrs,
                edge_feats=edge_feats,
                edge_index=data["edge_index"],
                cutoff=cutoff, 
            )
            node_feats = product(
                node_feats=node_feats,
                sc=sc,
                node_attrs=data["node_attrs"],
            )
            logging.info(f'node attribute change for layer {counter}')
            logging.info(node_feats)
            # node_out = readout(node_feats).squeeze(-1)
            node_feats_total.append(node_feats)
        
        node_out = node_feats_total[-1]

        inter_e = scatter_mean(
            src= node_out,
            index=data['batch'].unsqueeze(-1),
            dim=0,
            dim_size=num_graphs,
        )  # [n_graphs,16]
        inter_std = scatter_std(
            src=node_out,
            index=data['batch'].unsqueeze(-1),
            dim=0,
            dim_size=num_graphs,
        )  # [n_graphs,16]
        inter_sum = scatter_sum(
            src=node_out,
            index=data['batch'].unsqueeze(-1),
            dim=0,
            dim_size=num_graphs,
        )  # [n_graphs,16]

        inter_e = inter_e[:, :, None]
        inter_std = inter_std[:, :, None]
        inter_sum = inter_sum[:, :, None]

        graph_logits = self.readouts[0].forward((inter_e, inter_std, inter_sum))  # [n_graphs,1,16]
        node_feats = torch.cat([inter_e, inter_std, inter_sum], dim=2).flatten().tolist()
        probs     = torch.sigmoid(graph_logits).tolist()         # [batch_size]
        true_probs = data['coupling_class']

        output = {
            "coupling_class": graph_logits  # [n_nodes, n_classes]
        }

        return output
```
